### PR TITLE
Improve handling of types of semantics in glsl

### PR DESCRIPTION
**Pull request by jsmall-nvidia**
Attempt to improve the glsl handling of hlsl semantics by taking into account the underlying glsl type.

### CODE REVIEW

File: source/slang/ir-glsl-legalize.cpp [hunk 1]
```diff
@@ -285,80 +302,121 @@ GLSLSystemValueInfo* getGLSLSystemValueInfo(
...
     else if(semanticName == "sv_depthlessequal")
     {
         // TODO: layout(depth_greater) out float gl_FragDepth;
+
+        // 'unknown' in hlsl, float in glsl 
         name = "gl_FragDepth";
     }
     else if(semanticName == "sv_dispatchthreadid")
     {
+        // uint3 in hlsl, uvec3 in glsl
```
Comment (tfoleyNV) [inline source/slang/ir-glsl-legalize.cpp]: An important point that these comments leave out is that while there might be an "official" type for most of these in HLSL, in practice the user is allowed to declare almost anything that the HLSL compiler can implicitly convert to/from the correct type.

File: source/slang/ir-glsl-legalize.cpp [hunk 2]
```diff
@@ -285,80 +302,121 @@ GLSLSystemValueInfo* getGLSLSystemValueInfo(
...
+        // https://www.opengl.org/sdk/docs/manglsl/docbook4/xhtml/gl_GlobalInvocationID.xml
         name = "gl_GlobalInvocationID";
 
-        auto builder = context->getBuilder();
         requiredType = builder->getVectorType(builder->getBasicType(BaseType::UInt), builder->getIntValue(builder->getIntType(), 3));
     }
     else if(semanticName == "sv_domainlocation")
     {
+        // TODO: Not 100% confident that say float2 will convert into float3 glsl?
```
Comment (tfoleyNV) [inline source/slang/ir-glsl-legalize.cpp]: No, I would expect `adjustType` to have to make the change in vector width explicitly.

File: source/slang/ir-glsl-legalize.cpp [hunk 3]
```diff
@@ -376,27 +434,46 @@ GLSLSystemValueInfo* getGLSLSystemValueInfo(
...
+
+        requiredType = builder->getBasicType(BaseType::Int);
+
         context->requireGLSLExtension("ARB_shader_stencil_export");
         name = "gl_FragStencilRef";
     }
     else if (semanticName == "sv_tessfactor")
     {
+        // TODO(JS): Need to ensure the adjustType can handle such a scenario. 
+        // float[2|3|4] in hlsl, float[4] on glsl (ie both are arrays but might be different size)
```
Comment (tfoleyNV) [inline source/slang/ir-glsl-legalize.cpp]: Note also that this seems to be a case where HLSL nominally declares it as an array, but might allow a vector (or scalar) as well (I haven't looked into it).

File: source/slang/ir-glsl-legalize.cpp [hunk 4]
```diff
@@ -425,6 +502,16 @@ GLSLSystemValueInfo* getGLSLSystemValueInfo(
...
     {
+        // TODO: This doesn't seem to work correctly on it's own between hlsl/glsl
+
+        // Indeed on slang issue 109 claims this remains a problem  
+        // https://github.com/shader-slang/slang/issues/109
+
+        // On hlsl it's UINT related. "higher 16 bits for the right view, lower 16 bits for the left view."
+        // There is use in hlsl shader code as uint4 - not clear if that varies 
+        // https://github.com/KhronosGroup/GLSL/blob/master/extensions/nvx/GL_NVX_multiview_per_view_attributes.txt
+        // On glsl its highp int gl_ViewportMaskPerViewNV[];
```
Comment (tfoleyNV) [inline source/slang/ir-glsl-legalize.cpp]: Yeah, we originally had a request to support the NVIDIA GLSL extensions for multi-view, but when I started to dig into them it began to appear that there wasn't really a close match between the HLSL NVAPI extension and the GLSL extension, so the conversion wouldn't really be easy.

File: source/slang/ir-glsl-legalize.cpp (hunk 1 above)
Comment (jsmall-nvidia) [reply source/slang/ir-glsl-legalize.cpp]: So by this argument - does that imply we should always specify the glsl type, such that the type convert will be attempted for glsl output, and in doing so allowing more valid hlsl work with glsl output?

**Pull request merged**